test/parallel/test-readline-interface-recursive-writes.js

```diff
@@ -0,0 +1,30 @@
+'use strict';
+const common = require('../common');
+const ArrayStream = require('../common/arraystream');
+const assert = require('assert');
+
+common.skipIfDumbTerminal();
+
+const readline = require('readline');
+const rli = new readline.Interface({
+  terminal: true,
+  input: new ArrayStream(),
+});
+
+let recursionDepth = 0;
+
+rli.on('line', function onLine() {
+  // Abort in case of infinite loop
+  if (recursionDepth > 2) {
+    return;
+  }
+  recursionDepth++;
+  // Write something recursively to readline
+  rli.write('foo');
+});
+
+// Minimal reproduction for #46731
+const testInput = ' \n}\n';
+rli.write(testInput);
+
+assert.strictEqual(recursionDepth, testInput.match(/\n/g).length);
```

Comment:
You're using 2 as a magic number here, let's refactor that.
```suggestion
rli.once('line', common.mustCall(() => {
  rli.once('line', common.mustCall(() => {
    rli.on('line', common.mustNotCall());
  }));

  rli.write('foo');
}));

// Minimal reproduction for #46731
rli.write(' \n}\n');

```

or
```suggestion
// Minimal reproduction for #46731
const testInput = ' \n}\n';
const numberOfExpectedLines = testInput.match(/\n/g).length;

rli.on('line', () => {
  // Abort in case of infinite loop
  if (recursionDepth > numberOfExpectedLines) {
    return;
  }
  recursionDepth++;
  // Write something recursively to readline
  rli.write('foo');
});


rli.write(testInput);

assert.strictEqual(recursionDepth, numberOfExpectedLines);
```